### PR TITLE
predicted-actual, residual plots back-end

### PR DESCRIPTION
**Pull request by cdathuraliya**

### CODE REVIEW

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java [hunk 1]
```diff
@@ -151,6 +152,31 @@ public static ClassClassificationAndRegressionModelSummary generateRegressionMod
             predictedVsActual.setActual(Double.parseDouble(decimalFormat.format(scoreAndLabel._2())));
             predictedVsActuals.add(predictedVsActual);
         }
+        // create a list of feature values
+        List<double[]> features = new ArrayList<double[]>();
+        for (LabeledPoint labeledPoint : testingData.collect()) {
+            double[] rowFeatures = labeledPoint.features().toArray();
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java]: Null checks are missing.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java [hunk 2]
```diff
@@ -184,6 +211,31 @@ public static ClassClassificationAndRegressionModelSummary getClassClassificatio
             predictedVsActual.setActual(scoreAndLabel._2());
             predictedVsActuals.add(predictedVsActual);
         }
+        // create a list of feature values
+        List<double[]> features = new ArrayList<double[]>();
+        for (LabeledPoint labeledPoint : testingData.collect()) {
+            double[] rowFeatures = labeledPoint.features().toArray();
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java]: Null check.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java [hunk 3]
```diff
@@ -92,8 +92,13 @@ public static ProbabilisticClassificationModelSummary generateProbabilisticClass
...
         List<double[]> features = new ArrayList<double[]>();
         for (LabeledPoint labeledPoint : testingData.collect()) {
-            double[] rowFeatures = labeledPoint.features().toArray();
-            features.add(rowFeatures);
+            if(labeledPoint != null && labeledPoint.features() != null) {
+                double[] rowFeatures = labeledPoint.features().toArray();
+                features.add(rowFeatures);
+            }
+            else {
+                continue;
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java:100]: "continue" can be removed..?

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java [hunk 4]
```diff
@@ -151,6 +157,36 @@ public static ClassClassificationAndRegressionModelSummary generateRegressionMod
...
             predictedVsActuals.add(predictedVsActual);
         }
+        // create a list of feature values
+        List<double[]> features = new ArrayList<double[]>();
+        for (LabeledPoint labeledPoint : testingData.collect()) {
+            if(labeledPoint != null && labeledPoint.features() != null) {
+                double[] rowFeatures = labeledPoint.features().toArray();
+                features.add(rowFeatures);
+            }
+            else {
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java:167]: no need of else block

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java [hunk 5]
```diff
@@ -184,6 +221,36 @@ public static ClassClassificationAndRegressionModelSummary getClassClassificatio
...
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java:231]: no need of else block

Comment (cdathuraliya) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java:231]: Removed and updated.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java (hunk 3 above)
Comment (cdathuraliya) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java:100]: Removed and updated.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java (hunk 4 above)
Comment (cdathuraliya) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SparkModelUtils.java:167]: Removed and updated.

**Pull request merged**